PR for #64 -  maps

gabrieleborg opened the pull request:
Sending the maps to PR @diegogentilepassaro  @santiagohermo. You can look at my last edits [here](https://github.com/diegogentilepassaro/min_wage_rent/issues/64#issuecomment-724950422). 

An important thing you'll notice is that the script does not follow the classic setup, main(), programs() structure, but a linear one. The reason is that, for unknown reasons, when I wrap all the code in the main function maps are not rendered. In that case we end up with either empty png files, or containing the legend only. I spent some time researching why but I couldn't find a solution.

santiagohermo commented on analysis/descriptive/code/plot_maps.R:
```diff
@@ -0,0 +1,551 @@
+  sample_zcta <- zipzcta[sample_zip, on = .(zipcode)][, .(zcta)][, value := 1]
+  setnames(sample_zcta, old = 'zcta', new = 'region')
+  
+  data(zip.map)
```
Follow-up question: should we download those files and add them to the repo? Probably doesn't make much sense, since it's not our main focus to build maps

gabrieleborg replied:
Question 1: yes. I find it the easiest and fastest way to get geographical data at the zcta level for the whole US

Question 2: I thought about it, and we could do it. But at the same time they're very fast and easy to download so for the moment I left it like that.

santiagohermo replied:
Sounds good, thanks! I agree that, as long as it's fast and not central to our paper, requesting the data from the internet makes sense.

santiagohermo commented on analysis/descriptive/code/plot_maps.R:
```diff
@@ -0,0 +1,551 @@
+
+
+
+plot_changes_city <- function(target_var,
+                              target_msa, 
+                              mwarea, 
+                              mwdate, 
+                              plotname, 
+                              df_data, 
+                              nmon,
+                              zctamsa, 
+                              zipzcta,
+                              zctaplace,
+                              out) {
```
Just curious here. What is the logic of writing the arguments like this? I've seen this style a lot but I'm refusing to adopt it myself haha

gabrieleborg replied:
Just visual clarity I guess 😁

santiagohermo commented on analysis/descriptive/code/plot_maps.R:
```diff
@@ -0,0 +1,551 @@
+  return(plot)
+}
+
+plot_mw_changes_city <- function(target_var,
```
Do we need two functions, one for mw changes and other for rents? It sounds like we are only changing the variable but we still rewrite the whole structure.

gabrieleborg replied:
I started with 2 but I eventually converged on something very similar, you are right! I'll consolidate them

santiagohermo commented on analysis/descriptive/code/plot_maps.R:
```diff
@@ -0,0 +1,551 @@
+  df_target[, 'region' := zipcode]
+  df_target[, pct_rentch := round(Winsorize(pct_rentch)*100, digits = 2)] #winsorize at .05 and .95
+  
+  data(zip.map)
+  zcta_map <- sfheaders::sf_multipolygon(zip.map, x = 'long', y = 'lat', multipolygon_id = 'id', polygon_id = 'piece', keep = T)
```
If feels like this zcta_map object can be created once in the bottom and then passed as an argument to each function.

gabrieleborg replied:
good point

santiagohermo commented on analysis/descriptive/code/plot_maps.R:
```diff
@@ -0,0 +1,551 @@
+df <- setDT(df)
+df[, countyfips := str_pad(as.character(countyfips), 5, pad = 0)]
+df[, zipcode := str_pad(as.character(zipcode), 5, pad = 0)]
+df[, year_month := fromStataTime(year_month, '%tm')]
```
cool!

santiagohermo commented on analysis/descriptive/code/plot_maps.R:
```diff
@@ -0,0 +1,551 @@
+# L.A. MSA
+png(filename = paste0(outdir, 'Los_Angeles_msa.png'), width = 7680, height = 7680)
+plot_changes_city(target_var = 'medrentpricepsqft_sfcc', 
+                target_msa = "Los Angeles",
+                nmon = 6,
+                plotname = "Los Angeles MSA",
+                df_data = df,
+                mwarea = 44000,
+                mwdate = '2019-07-01',
+                zctamsa = zcta_msa_xwalk,
+                zipzcta = zip_zcta_xwalk,
+                zctaplace = zcta_place_xwalk,
+                out = outdir)
+dev.off()
+png(filename = paste0(outdir, 'Los_Angeles_msa_mw.png'), width = 7680, height = 7680)
+plot_mw_changes_city(target_var = 'actual_mw', 
+                 target_msa = "Los Angeles",
+                 nmon = 6,
+                 plotname = "Los Angeles MSA",
+                 df_data = df,
+                 mwarea = 44000,
+                 mwdate = '2019-07-01',
+                 zctamsa = zcta_msa_xwalk,
+                 zipzcta = zip_zcta_xwalk,
+                 zctaplace = zcta_place_xwalk,
+                 out = outdir)
+dev.off()
+png(filename = paste0(outdir, 'Los_Angeles_msa_expmw.png'), width = 7680, height = 7680)
+plot_mw_changes_city(target_var = 'exp_mw_totjob', 
+                     target_msa = "Los Angeles",
+                     nmon = 6,
+                     plotname = "Los Angeles MSA",
+                     df_data = df,
+                     mwarea = 44000,
+                     mwdate = '2019-07-01',
+                     zctamsa = zcta_msa_xwalk,
+                     zipzcta = zip_zcta_xwalk,
+                     zctaplace = zcta_place_xwalk,
+                     out = outdir)
+dev.off()
```
A nice `lapply` here will probably make this shorter. Something like

```r
suffix <- c("", "_mw", "_expmw")
vars <- c("medrentpricepsqft_sfcc", "actual_mw", "exp_mw_totjob")

lapply(1:3, function (x) {
   png(filename = paste0(outdir, 'Los_Angeles_msa', suffix[i], '.png'), width = 7680, height = 7680)
   plot_mw_changes_city(target_var = vars[i], 
                     target_msa = "Los Angeles",
                     nmon = 6,
                     plotname = "Los Angeles MSA",
                     df_data = df,
                     mwarea = 44000,
                     mwdate = '2019-07-01',
                     zctamsa = zcta_msa_xwalk,
                     zipzcta = zip_zcta_xwalk,
                     zctaplace = zcta_place_xwalk,
                     out = outdir)
   dev.off()
   }
)
``` 

You could also loop across some vectors that store MSA names, and define all MSAs you want in one place. This would make it easier to change what maps you want to build.

santiagohermo reviewed (approved):
I completed a high level review of the code @gabrieleborg. I made some questions just to make sure I understand and made a few suggestions to simplify the code a bit. Once thing to flag is that this script doesn't follow our logic of having a main program and defining parameters and stuff within it. I guess that we may want to change at some point, but no need to do that now as we have more pressing issues!

Overall it looks great, thanks Gabri! It seems to me that it does what we want quite well. I'm approving, so that once you are done with the improvements we discussed today you just go ahead and merge.

Also, I updated this branch from `master` and from branch `#67`. Now you have the output tex in staged output ([this one](https://github.com/diegogentilepassaro/min_wage_rent/blob/75b8d2f7923dde9696ae1cd490b46f685ddfcfc1/paper_slides/staged_output/output.tex)). You can construct the figure that we want to add to the paper there, so that once we decide how to put it we just copy-paste, and maybe add some notes and stuff.

gabrieleborg commented on analysis/descriptive/code/plot_maps.R:
```diff
@@ -0,0 +1,551 @@
+                      legend_name,
+                      zctamsa, 
+                      out) {
+  census_api_key('6e4bf831630a20872606ce3949acd975473a1d87')
+  #censusvar <- load_variables(year = 2010, dataset = 'sf1') #to navigate census variables
+  census_target <- c(pop = 'P001001', urbpop = 'P002002', housing_units = 'H001001', house_urban = 'H002002')
+  df_census <- get_decennial(geography = 'zcta', year = 2010, state = NULL, geometry = F,
+                             variables = census_target, output = 'wide') %>%
+    rename('zcta' = 'GEOID') %>%
+    select(- 'NAME') %>%
+    setDT()
```
Yes, we are getting the few variables from the Census to plot the density! I am going to simplify this because: 

1. the total/urban distinction doesn't matter since we are plotting info only for MSA areas. 
2. the populaiton- and house-based maps look identical

gabrieleborg commented:
Thanks @santiagohermo for the comments. Regarding your main comments [here](https://github.com/diegogentilepassaro/min_wage_rent/pull/69#pullrequestreview-529528572), please see my first post on why I did not apply the classic "main() function" structure. I will implement the changes you suggested and merge soon!

santiagohermo commented:
Thanks @gabrieleborg! Sorry for missing your point in the opening comment. I think the linear structure is ok for now, we can revisit the pipeline after our deadline. 

I looked at the changes you implemented thus far and they look good. One quick comment about the legend. I think it should either be "6-month change" or "6 months change".

gabrieleborg commented:
@santiagohermo I implemented some final changes addressing the remaining issues: 

1. I managed to restructure the script and use the main() function structure. I just needed to wrap the plot in a print() function. 
2. I re-structured the city plot function that now works with a lapply() as you suggested. 
3. I removed the title and subtitle (reporting the MW change) from the figure because they didn't look good in the final latex. We can report the MSA name in the figure title and the MW date change in the figure notes. 

I am going on and merging the branch!

santiagohermo commented:
Great @gabrieleborg! I'll also delete the branch
